File: backend/swagger/api-client-example.js
#!/usr/bin/env node

/**
 * Web Appliance Dashboard API Client Example
 * 
 * This script demonstrates how to use the Web Appliance Dashboard API with Node.js.
 * It includes examples for all major endpoints with proper error handling.
 * 
 * Requirements:
 *   npm install node-fetch
 * 
 * Usage:
 *   node api-client-example.js
 */

const fetch = require('node-fetch');

class WebApplianceAPIClient {
  constructor(baseUrl = 'http://localhost:9080') {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.token = null;
  }

  /**
   * Handle API response and throw errors if needed
   */
  async handleResponse(response) {
    if (!response.ok) {
      let errorMessage = `HTTP Error ${response.status}`;
      try {
        const errorData = await response.json();
        errorMessage = `${errorMessage}: ${errorData.error || 'Unknown error'}`;
      } catch (e) {
        // Response might not be JSON
      }
      throw new Error(errorMessage);
    }
    
    const contentType = response.headers.get('content-type');
    if (contentType && contentType.includes('application/json')) {
      return response.json();
    }
    return response;
  }

  /**
   * Make an API request with authentication
   */
  async apiRequest(endpoint, options = {}) {
    const headers = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    if (this.token && !endpoint.includes('/auth/login')) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    const response = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers,
    });

    return this.handleResponse(response);
  }

  // Authentication
  async login(username, password) {
    const data = await this.apiRequest('/api/auth/login', {
      method: 'POST',
      body: JSON.stringify({ username, password }),
    });
    
    this.token = data.token;

    return data;
  }

  // Appliances
  async getAppliances() {
    return this.apiRequest('/api/appliances');
  }

  async createAppliance(applianceData) {
    return this.apiRequest('/api/appliances', {
      method: 'POST',
      body: JSON.stringify(applianceData),
    });
  }

  async updateAppliance(id, updates) {
    return this.apiRequest(`/api/appliances/${id}`, {
      method: 'PUT',
      body: JSON.stringify(updates),
    });
  }

  async deleteAppliance(id) {
    return this.apiRequest(`/api/appliances/${id}`, {
      method: 'DELETE',
    });
  }

  // Categories
  async getCategories() {
    return this.apiRequest('/api/categories');
  }

  // Services
  async getServices() {
    return this.apiRequest('/api/services');
  }

  async controlService(serviceName, action) {
    return this.apiRequest(`/api/services/${serviceName}/${action}`, {
      method: 'POST',
    });
  }

  // Settings
  async getSettings() {
    return this.apiRequest('/api/settings');
  }

  async updateSetting(key, value) {
    return this.apiRequest(`/api/settings/${key}`, {
      method: 'PUT',
      body: JSON.stringify({ value }),
    });
  }

  // SSH Keys
  async getSSHKeys() {
    return this.apiRequest('/api/ssh/keys');
  }

  async generateSSHKey(name, passphrase = '') {
    return this.apiRequest('/api/ssh/keys/generate', {
      method: 'POST',
      body: JSON.stringify({ name, passphrase }),
    });
  }

  // Status Check
  async checkApplianceStatus(applianceIds) {
    return this.apiRequest('/api/statusCheck', {
      method: 'POST',
      body: JSON.stringify({ applianceIds }),
    });
  }

  // Audit Logs
  async getAuditLogs(options = {}) {
    const params = new URLSearchParams({
      limit: options.limit || 100,
      offset: options.offset || 0,
      ...(options.action && { action: options.action }),
      ...(options.userId && { userId: options.userId }),
    });

    return this.apiRequest(`/api/auditLogs?${params}`);
  }

  // Backup
  async createBackup(outputFile) {
    const fs = require('fs');
    const response = await fetch(`${this.baseUrl}/api/backup`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Backup failed: ${response.statusText}`);
    }

    const buffer = await response.buffer();
    fs.writeFileSync(outputFile, buffer);

  }

  // Monitoring
  async monitorAppliances(interval = 60000, duration = null) {
    const startTime = Date.now();

    const checkStatus = async () => {
      try {
        const appliances = await this.getAppliances();
        if (!appliances.length) {

          return;
        }

        const applianceIds = appliances.map(app => app.id);
        const statuses = await this.checkApplianceStatus(applianceIds);

        appliances.forEach(app => {
          const statusInfo = statuses[app.id] || {};
          const status = statusInfo.status || 'unknown';
          const responseTime = statusInfo.responseTime || '-';
          
          const statusIcon = status === 'online' ? '🟢' : 
                           status === 'offline' ? '🔴' : '⚪';

        });

        if (duration && (Date.now() - startTime) >= duration) {

          clearInterval(intervalId);
        }
      } catch (error) {
        console.error('Monitor error:', error.message);
      }
    };

    // Initial check
    await checkStatus();

    // Set up interval
    const intervalId = setInterval(checkStatus, interval);

    // Handle Ctrl+C
    process.on('SIGINT', () => {

      clearInterval(intervalId);
      process.exit(0);
    });

    return intervalId;
  }

  // Server-Sent Events
  subscribeToUpdates(onUpdate, onError) {
    const EventSource = require('eventsource');
    const eventSource = new EventSource(
      `${this.baseUrl}/api/sse?token=${this.token}`
    );

    eventSource.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        onUpdate(data);
      } catch (error) {
        console.error('Failed to parse SSE data:', error);
      }
    };

    eventSource.onerror = (error) => {
      onError(error);
      eventSource.close();
    };

    return () => eventSource.close();
  }
}

// Main demonstration function
async function main() {
  const client = new WebApplianceAPIClient();

  try {
    // 1. Login

    await client.login('admin', 'password123');

    // 2. Get and display appliances

    const appliances = await client.getAppliances();
    appliances.forEach(app => {

    });

    // 3. Create a new appliance

    const newAppliance = await client.createAppliance({
      name: 'Example Application',
      url: 'http://example.local:8080',
      description: 'Example application for testing',
      icon: 'Globe',
      color: '#FF5733',
      category: 'development',
    });

    // 4. Update the appliance

    await client.updateAppliance(newAppliance.id, {
      description: 'Updated description for example app',
    });

    // 5. Check appliance status

    const statuses = await client.checkApplianceStatus([newAppliance.id]);
    const statusInfo = statuses[newAppliance.id] || {};

    // 6. Get categories

    const categories = await client.getCategories();
    categories.forEach(cat => {

    });

    // 7. Get settings

    const settings = await client.getSettings();
    const settingsMap = Object.fromEntries(
      settings.map(s => [s.key, s.value])
    );

    // 8. Get audit logs

    const logs = await client.getAuditLogs({ limit: 5 });
    logs.logs.slice(0, 5).forEach(log => {

    });

    // 9. Subscribe to real-time updates

    const unsubscribe = client.subscribeToUpdates(
      (data) => console.log('Update:', data),
      (error) => console.error('SSE Error:', error)
    );

    // Stop after 10 seconds
    setTimeout(() => {
      unsubscribe();

    }, 10000);

    // 10. Monitor appliances for 30 seconds

    await client.monitorAppliances(10000, 30000);

    // 11. Clean up - delete the test appliance

    await client.deleteAppliance(newAppliance.id);

  } catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    process.exit(1);
  }
}

// Run the example if this is the main module
if (require.main === module) {
  main().catch(console.error);
}

module.exports = WebApplianceAPIClient;

File: backend/swagger/api-client-example.py
#!/usr/bin/env python3
"""
Web Appliance Dashboard API Client Example

This script demonstrates how to use the Web Appliance Dashboard API with Python.
It includes examples for all major endpoints with proper error handling.

Requirements:
    pip install requests

Usage:
    python api-client-example.py
"""

import requests
import json
import time
from datetime import datetime
from typing import Dict, List, Optional, Any


class WebApplianceAPIClient:
    """Client for interacting with the Web Appliance Dashboard API."""
    
    def __init__(self, base_url: str = "http://localhost:9080"):
        """
        Initialize the API client.
        
        Args:
            base_url: The base URL of the API server
        """
        self.base_url = base_url.rstrip('/')
        self.token: Optional[str] = None
        self.session = requests.Session()
        
    def _get_headers(self) -> Dict[str, str]:
        """Get headers with authentication token if available."""
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        return headers
    
    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        """Handle API response and raise exceptions for errors."""
        try:
            response.raise_for_status()
            return response.json() if response.content else {}
        except requests.exceptions.HTTPError as e:
            error_msg = f"HTTP Error {response.status_code}"
            try:
                error_data = response.json()
                error_msg = f"{error_msg}: {error_data.get('error', 'Unknown error')}"
            except json.JSONDecodeError:
                pass
            raise Exception(error_msg) from e
        except requests.exceptions.RequestException as e:
            raise Exception(f"Request failed: {str(e)}") from e
    
    # Authentication
    def login(self, username: str, password: str) -> Dict[str, Any]:
        """
        Login to the API and store the authentication token.
        
        Args:
            username: Username for authentication
            password: Password for authentication
            
        Returns:
            User information and token
        """
        response = self.session.post(
            f"{self.base_url}/api/auth/login",
            json={"username": username, "password": password}
        )
        data = self._handle_response(response)
        self.token = data.get("token")
        print(f"✓ Logged in as {data['user']['username']}")
        return data
    
    # Appliances
    def get_appliances(self) -> List[Dict[str, Any]]:
        """Get all appliances."""
        response = self.session.get(
            f"{self.base_url}/api/appliances",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    def create_appliance(self, appliance_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new appliance."""
        response = self.session.post(
            f"{self.base_url}/api/appliances",
            headers=self._get_headers(),
            json=appliance_data
        )
        return self._handle_response(response)
    
    def update_appliance(self, appliance_id: int, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update an existing appliance."""
        response = self.session.put(
            f"{self.base_url}/api/appliances/{appliance_id}",
            headers=self._get_headers(),
            json=updates
        )
        return self._handle_response(response)
    
    def delete_appliance(self, appliance_id: int) -> Dict[str, Any]:
        """Delete an appliance."""
        response = self.session.delete(
            f"{self.base_url}/api/appliances/{appliance_id}",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    # Categories
    def get_categories(self) -> List[Dict[str, Any]]:
        """Get all categories with appliance counts."""
        response = self.session.get(
            f"{self.base_url}/api/categories",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    # Services
    def get_services(self) -> List[Dict[str, Any]]:
        """Get all system services."""
        response = self.session.get(
            f"{self.base_url}/api/services",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    def control_service(self, service_name: str, action: str) -> Dict[str, Any]:
        """
        Control a system service.
        
        Args:
            service_name: Name of the service
            action: Action to perform (start, stop, restart)
        """
        response = self.session.post(
            f"{self.base_url}/api/services/{service_name}/{action}",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    # Settings
    def get_settings(self) -> List[Dict[str, Any]]:
        """Get all system settings."""
        response = self.session.get(
            f"{self.base_url}/api/settings",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    def update_setting(self, key: str, value: str) -> Dict[str, Any]:
        """Update a system setting."""
        response = self.session.put(
            f"{self.base_url}/api/settings/{key}",
            headers=self._get_headers(),
            json={"value": value}
        )
        return self._handle_response(response)
    
    # SSH Keys
    def get_ssh_keys(self) -> List[Dict[str, Any]]:
        """Get all SSH keys."""
        response = self.session.get(
            f"{self.base_url}/api/ssh/keys",
            headers=self._get_headers()
        )
        return self._handle_response(response)
    
    def generate_ssh_key(self, name: str, passphrase: str = "") -> Dict[str, Any]:
        """Generate a new SSH key pair."""
        response = self.session.post(
            f"{self.base_url}/api/ssh/keys/generate",
            headers=self._get_headers(),
            json={"name": name, "passphrase": passphrase}
        )
        return self._handle_response(response)
    
    # Status Check
    def check_appliance_status(self, appliance_ids: List[int]) -> Dict[str, Any]:
        """Check status of multiple appliances."""
        response = self.session.post(
            f"{self.base_url}/api/statusCheck",
            headers=self._get_headers(),
            json={"applianceIds": appliance_ids}
        )
        return self._handle_response(response)
    
    # Audit Logs
    def get_audit_logs(self, limit: int = 100, offset: int = 0, 
                      action: Optional[str] = None, user_id: Optional[int] = None) -> Dict[str, Any]:
        """Get audit logs with optional filtering."""
        params = {"limit": limit, "offset": offset}
        if action:
            params["action"] = action
        if user_id:
            params["userId"] = user_id
            
        response = self.session.get(
            f"{self.base_url}/api/auditLogs",
            headers=self._get_headers(),
            params=params
        )
        return self._handle_response(response)
    
    # Backup
    def create_backup(self, output_file: str) -> None:
        """Create a system backup and save to file."""
        response = self.session.post(
            f"{self.base_url}/api/backup",
            headers=self._get_headers(),
            stream=True
        )
        response.raise_for_status()
        
        with open(output_file, 'wb') as f:
            for chunk in response.iter_content(chunk_size=8192):
                f.write(chunk)
        print(f"✓ Backup saved to {output_file}")
    
    # Monitoring
    def monitor_appliances(self, interval: int = 60, duration: Optional[int] = None):
        """
        Monitor appliance status continuously.
        
        Args:
            interval: Check interval in seconds
            duration: Total monitoring duration in seconds (None for infinite)
        """
        start_time = time.time()
        print(f"Starting appliance monitoring (interval: {interval}s)...")
        
        try:
            while True:
                # Get all appliances
                appliances = self.get_appliances()
                if not appliances:
                    print("No appliances configured")
                    break
                
                # Check status
                appliance_ids = [app['id'] for app in appliances]
                statuses = self.check_appliance_status(appliance_ids)
                
                # Display results
                print(f"\n[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}]")
                for app in appliances:
                    status_info = statuses.get(str(app['id']), {})
                    status = status_info.get('status', 'unknown')
                    response_time = status_info.get('responseTime', '-')
                    
                    status_icon = "🟢" if status == "online" else "🔴" if status == "offline" else "⚪"
                    print(f"{status_icon} {app['name']:<30} {status:<10} {response_time}ms")
                
                # Check if we should stop
                if duration and (time.time() - start_time) >= duration:
                    break
                
                # Wait for next check
                time.sleep(interval)
                
        except KeyboardInterrupt:
            print("\n\nMonitoring stopped by user")


def main():
    """Main function demonstrating API usage."""
    # Initialize client
    client = WebApplianceAPIClient()
    
    try:
        # 1. Login
        print("=== Authentication ===")
        client.login("admin", "password123")
        
        # 2. Get and display appliances
        print("\n=== Current Appliances ===")
        appliances = client.get_appliances()
        for app in appliances:
            print(f"- {app['name']} ({app['url']})")
        
        # 3. Create a new appliance
        print("\n=== Creating New Appliance ===")
        new_appliance = client.create_appliance({
            "name": "Example Application",
            "url": "http://example.local:8080",
            "description": "Example application for testing",
            "icon": "Globe",
            "color": "#FF5733",
            "category": "development"
        })
        print(f"✓ Created appliance: {new_appliance['name']} (ID: {new_appliance['id']})")
        
        # 4. Update the appliance
        print("\n=== Updating Appliance ===")
        updated = client.update_appliance(new_appliance['id'], {
            "description": "Updated description for example app"
        })
        print(f"✓ Updated appliance description")
        
        # 5. Check appliance status
        print("\n=== Checking Status ===")
        statuses = client.check_appliance_status([new_appliance['id']])
        status_info = statuses.get(str(new_appliance['id']), {})
        print(f"Status: {status_info.get('status', 'unknown')}")
        
        # 6. Get categories
        print("\n=== Categories ===")
        categories = client.get_categories()
        for cat in categories:
            print(f"- {cat['display_name']}: {cat['applianceCount']} appliances")
        
        # 7. Get settings
        print("\n=== Settings ===")
        settings = client.get_settings()
        settings_dict = {s['key']: s['value'] for s in settings}
        print(f"SSH Enabled: {settings_dict.get('ssh_enabled', 'N/A')}")
        print(f"Terminal Enabled: {settings_dict.get('terminal_enabled', 'N/A')}")
        
        # 8. Get audit logs
        print("\n=== Recent Audit Logs ===")
        logs = client.get_audit_logs(limit=5)
        for log in logs['logs'][:5]:
            print(f"- [{log['timestamp']}] {log['username']}: {log['action']}")
        
        # 9. Monitor appliances for 30 seconds
        print("\n=== Monitoring Appliances ===")
        client.monitor_appliances(interval=10, duration=30)
        
        # 10. Clean up - delete the test appliance
        print("\n=== Cleanup ===")
        client.delete_appliance(new_appliance['id'])
        print(f"✓ Deleted test appliance")
        
    except Exception as e:
        print(f"\n❌ Error: {e}")
        return 1
    
    return 0


if __name__ == "__main__":
    exit(main())
